src/vsr/routing.zig

```diff
@@ -0,0 +1,877 @@
+        };
+    }
+};
+
+const Routing = @This();
+
+pub fn init(options: struct {
+    replica: u8,
+    replica_count: u8,
+    standby_count: u8,
+}) Routing {
+    assert(options.replica < options.replica_count + options.standby_count);
+    assert(options.replica_count <= constants.replicas_max);
+    assert(options.standby_count <= constants.standbys_max);
+
+    const route = route_view_default(0, options.replica_count);
+
+    return .{
+        .replica = options.replica,
+        .replica_count = options.replica_count,
+        .standby_count = options.standby_count,
+
+        .view = 0,
+        .a = route,
+        .a_cost = null,
+
+        .b = route,
+        .b_cost = null,
+
+        .history = .{OpHistory.root()} ** history_max,
+    };
+}
+
+pub fn view_change(routing: *Routing, view: u32) void {
+    assert(view > routing.view or (view == 0 and routing.view == 0));
+    assert(routing.history_empty());
+    routing.view = view;
+
+    const route = route_view_default(view, routing.replica_count);
+    assert(routing.route_valid(route));
+
+    routing.a = route;
+    assert(routing.a_cost == null);
+
+    routing.b = route;
+    assert(routing.b_cost == null);
+}
+
+fn route_valid(routing: *const Routing, route: Route) bool {
+    if (route.count() != routing.replica_count) return false;
+
+    for (route.const_slice()) |replica| {
+        if (replica >= routing.replica_count) return false;
+    }
+    for (0..routing.replica_count) |i| {
+        for (0..i) |j| {
+            if (route.get(i) == route.get(j)) return false;
+        }
+    }
+
+    const primary: u8 = @intCast(routing.view % routing.replica_count);
+    const primary_index = @divFloor(route.count(), 2);
+    if (route.get(primary_index) != primary) return false;
+
+    return true;
+}
+
+fn route_view_default(view: u32, replica_count: u8) Route {
+    var route: Route = .{};
+    for (0..replica_count) |replica| {
+        route.append_assume_capacity(@intCast(replica));
+    }
+
+    // Rotate primary to the midpoint;
+    const primary: u8 = @intCast(view % replica_count);
+    const midpoint = @divFloor(replica_count, 2);
+    const rotation = (replica_count + primary - midpoint) % replica_count;
+    std.mem.rotate(u8, route.slice(), rotation);
+    assert(route.slice()[midpoint] == primary);
+
+    return route;
+}
+
+fn route_random(prng: *stdx.PRNG, view: u32, replica_count: u8) Route {
+    var route: Route = .{};
+    for (0..replica_count) |replica| {
+        route.append_assume_capacity(@intCast(replica));
+    }
+    prng.shuffle(u8, route.slice());
+
+    const primary: u8 = @intCast(view % replica_count);
+    const primary_index = std.mem.indexOfScalar(u8, route.slice(), primary).?;
+    std.mem.swap(
+        u8,
+        &route.slice()[primary_index],
+        &route.slice()[@divFloor(replica_count, 2)],
+    );
+
+    return route;
+}
+
+pub fn route_encode(routing: *const Routing, route: Route) u64 {
```

Comment:
Originally I went for a crazy scheme that encoded a route in 255 as an index of permutation of `replica_count - 1` replicas. That was fancy enough to warrant a blog-post: https://matklad.github.io/2025/04/15/underusing-snapshot-testing.html. 

So I reconsidered and went for direct encoding of a list of replicas.